### DISCUSSION TITLE
Fake spd_kt using GNSS data

### DISCUSSION THREAD

**Discussion by MartinPersson** [Rodemfr/MicronetToNMEA#77] [Ideas]
A great feature (at least to me) would be to feed spd_kt data to the micronet system without having access to VHW NMEA sentences. To my understanding, this, together with magHdg_deg, would allow VMG to wind and True wind to be displayed on an MN100 also in the absence of VHW data or a T121 Hull Transmitter. Using speed data from the VTG sentences should be a good replacement for actual speed through water, and even more accurate in the presence of strong currents.

**Comment by dwarning** [comment]
What would be the source of the VTG data, from GNSS? The attached Matrix is a trial to structure the data flow, perhaps more from my configuration with hull transmitter. Can you clarify the data flow in your boat configuration?
[NMEA_Matrix.xls](https://github.com/Rodemfr/MicronetToNMEA/files/10838395/NMEA_Matrix.xls)

**Comment by tvr256** [comment]
True Wind is wind speed/direction over water, whereas Ground Wind is wind speed/direction over ground.

To calculate True Wind, you need to know the boat speed through water, which requires a water speed transducer.

It sounds like @MartinPersson wants to use Ground Wind / COG / SOG when no water speed transducer is fitted, instead of True Wind / STW / BTW.

Some interesting reading here:

https://raymarine.custhelp.com/app/answers/detail/a_id/3794/~/apparent-wind%2C-true-wind-and-ground-wind%2C-and-data-required-to-calculate-them

https://forums.ybw.com/index.php?threads/paddlewheel-simulator-boat-speed-vs-sog-ground-wind-log.567653/

**Comment by Rodemfr** [comment]
We could add a new compile switch in `BoardConfig.h` to enable this feature. Something like
```
#define SPD_FROM_SOG 1
```
It is pretty easy. I'm however wondering if we would need a piece of code to disable this conversion in case `spd_kt` would be received anyway from the transducer. I could imaging someone with this flag enabled to later add a speed transducer to his system and not realizing that the speed displayed is the one from GNSS, not the one from is new transducer.

**Reply by tvr256** [reply]
Weirdly, I think that's exactly what my older TackTick display does! See my previous issue #72 , older Tacktick firmware derives STW from SOG if no speed transducer is detected.

Newer Tacktick firmware displays null water speed instead.

**Reply by MartinPersson** [reply]
Really? That’s great news (although referred to as a bug and a bad idea in issue #72, hehe). I have FW 21 on mine, so even older than yours. I guess I’ll see if have  have this feature/bug once I get the MTN device up and running. Waiting for hardware right now.

**Comment by MartinPersson** [comment]
First of all, thank you for a great project (missed this in my original post)!
Second of all, thank you both for taking the time to reply. 
Third of all, feel slightly embarrassed for not knowing that True Wind is defined as wind speed/direction over water. I thought it was calculated this way because the Tacktick systems came out before most people had a GPS onboard. Thanks for the links @tvr256.

Anyhow, my setup is currently this:
Tacktick T120 Wind Transducer
Tacktick T112 Analogue Display
Tacktick T111 Dual Display
MicronetToNMEA device with GNSS and compass modules (to be added)

The data flow would I imagined would be something like this:
-Wind data taken from T120 and only used by Micronet devices.
-GNSS data (VTG etc.) and magnetic compass data taken from the modules attached to the MicronetToNMEA device and translated to Micronet. 
-Derived units (specifically VMG to wind) calculated by the Tacktick devices, but using SOG disguised as STW.

I also have a Hull transmitter and a Wireless interface with defect batteries, and depth and speed transducers which are not mounted. 
I thought MicronetToNMEA running on a device with GNSS and compass attached could be an alternative to reviving/mounting those. I realize now that it won't give actual VMG to wind and True wind, but as I only sail in areas with no or very little current and no tides, interpreting SOG as STW should be close enough.

**Reply by tvr256** [reply]
MTN will replace the the T122 NMEA interface, the external GPS, and the external compass. But you'll still need a separate depth sender and water speed sensor.

You can either use analogue depth and wind sensors wired to the T121 Hull sender, or you can use NMEA sensors and send the data to the MTN NMEA_EXT input.

**Reply by tvr256** [reply]
The batteries on my T121 NMEA interface and T122 Hull transmitter are pretty ancient so they only hold charge for a minute or so. But it's not a problem, they're powered by 12v anyway. The only reason for the TackTick batteries is so you can switch off the boat electrics while the engine isn't running. But the TackTick uses so little power that leaving everything connected isn't an issue.

**Reply by MartinPersson** [reply]
I have an ancient Silva depth sounder that works reliably so far. Might replace it with one outputting NMEA and connect it to the MTN once it fails. 

From what I’ve gathered on the Raymarine support forum, the units won’t power on even with external power if the batteries are completely dead. At least I’m hoping that’s what’s wrong with mine. I’ll try to fix them by replacing the batteries although I’ll probably use the MTN instead anyways. Opening up the T122 (old white one) I found that the batteries are soldered directly onto the inward facing side of the PCB, which is in turn soldered to the screw terminals going though the case. So it can’t be removed without desoldering all the terminals. These things were obviously not meant to be serviced.

**Reply by tvr256** [reply]
Unfortunately you have the very early units without replaceable batteries. On the newer ones there's a removable battery cover...

**Comment by Rodemfr** [comment]
Your request makes sense Martin. I will add it soon to MTN. There are many reasons for not having water speed measurements in a boat. I will not make this features available by default to avoid having users reading SOG speed instead of water speed without knowing. 
A compile switch will make the trick.

Concerning the dead batteries, I changed some of my old batteries with new ones by ordering them at a good price on this website : https://www.laboiteapiles.com/produit/recherche?search=raymarine
It is in French but I imagine that you can find an equivalent site in English. I successfully changed the batteries of my T121, T111 and T120. Batteries were removable. Only T120 (wind transducer) was tricky to open.

**Reply by MartinPersson** [reply]
Great! Looking forward to testing it. 

I’ve actually just placed an order from that place with some help from google translate. Very good price compared to the official battery kits. Do you have the old white T121 or a newer one? I think the gray ones were redesigned to simplify battery replacement.

**Reply by Rodemfr** [reply]
I have both : a new gray one and a "old" white one. I changed batteries on my white T121 successfully without difficulty. Compared to newer versions, the only difference I found is that batteries where glued to the plastic hull.
By the way, if you manage to repair your T122, it would be very interesting to analyze its RF commands to catch a few remaining commands : magnetic variation comming from NMEA and performance data for T110.

**Reply by tvr256** [reply]
@Rodemfr , did your white one have a battery compartment cover or did you have to completely dismantle the entire unit down to the PCB? None of my (ancient) TackTick units have a battery cover. I tried opening up the NMEA sender, but it's impossible to get to the battery without unsoldering the connector plugs from the main PCB. As far as I can work out, this was only an issue on the very early white ones, and they fixed it with the later white ones and the grey ones.

**Reply by Rodemfr** [reply]
Mine has a battery compartment cover, like the gray ones. From what you wrote, it seems that I have a more recent white T121 than you. I will take a picture tonight to show you. Maybe is there is a manufacture date on a sticker on the back.

**Reply by tvr256** [reply]
Here's a comparison between my old and new T111 dual displays. As you can see, they didn't have a battery compartment in September 2005.

![image](https://user-images.githubusercontent.com/117984987/221834543-26a5f6f2-a685-4000-9d2d-292a38b8e6e5.png)

**Reply by Rodemfr** [reply]
Here is mine :
![Image](https://user-images.githubusercontent.com/62835095/221953227-514c1c52-6489-43c3-a09a-bdbd03d3264d.JPG)
It is obviously earlier. My FW version is 29.

**Reply by MartinPersson** [reply]
Here are my T121 and T122. Both marked 01/2005. 
![216AFFE9-0F58-430D-A387-4EF3F4CABA0A](https://user-images.githubusercontent.com/533128/221967526-ddbaa5f0-3907-4cb4-8150-449027101d67.jpeg)

I can see the batteries underneath the PCBs. Just like @tvr256, I’ll have to unsolder the connectors to remove the PCB and then unsolder the batteries from the PCB to replace them. 
Noted that they’ve used the same PCB for both, only rotated 90 deg and with a few components being different. Just a minor curiosity.

**Reply by MartinPersson** [reply]
I’d be happy to help catching the remaining commands if I can, @Rodemfr. Will ping you when I have the MTN up and running so you can explain what I need to do in more detail.

**Reply by Rodemfr** [reply]
Good luck in changing batteries. Let me know if you manage to recover your T122. The commands to test should not be to hard to setup.

**Comment by Rodemfr** [comment]
I implemented SPD emulation in branch [add_spd_emulation](https://github.com/Rodemfr/MicronetToNMEA/tree/add_spd_emulation). Can you test it with your system ?

**Reply by tvr256** [reply]
I've just tried it and it doesn't seem to be working like I'd expect...

With just the MTN powered up (and no TackTick components connected), no VHW sentence is sent out. I'd expect the VHW sentence to be sent regardless of whether the display is connected?

With the TackTick display connected (but no hull sensor or speed transducer), MTN now sends out a VHW sentence `$INVHW,130.2,T,129.2,M,0.8,N,,K*7c`. But the TackTick display still doesn't show the speed on the screen:

![PXL_20230302_114813662](https://user-images.githubusercontent.com/117984987/222420894-7ef03dab-1489-4159-81dd-f76d8a1a0e16.jpg)

**Comment by Rodemfr** [comment]
Thanks for the feedback.

For your first remark, I think it is "normal" in that sense that all non GNSS related sentences are output when the Micronet Master (the display) sends a message. I will see if it can be changed without too much pain to allow VHW to be sent without display.

For the second remark, I will test this tonight and find the bug...

**Reply by tvr256** [reply]
For the first point, I had assumed you were faking the speed signal by sending out a fake VHW sentence alongside the real VTG sentence. But looking at your code, you're being cleverer than that!

**Reply by Rodemfr** [reply]
I modified the branch : now using `EMULATE_SPD_WITH_SOG` switch will *always* copy SOG into SPD and it should be properly displayed by your T111 now.

**Comment by Rodemfr** [comment]
I released MicronetToNMEA v2.1 with SPD emulation code (disabled by default in BoardConfig.h)

**Comment by tvr256** [comment]
Works perfectly now, thanks!

With a wind sensor attached and SPD emulation disabled, I see an MWV (relative) sentence. With SPD emulation enabled, I also see MWV (true) and VHW sentences, and the SPEED value on the display matches the SOG value.

### Emulate disabled:

TackTick Display Off:

```
$INHDG,132.8,0,E,1.0,E*59
$GNRMC,141948.00,A,5115.59932,N,00040.68149,W,0.654,,060323,,,A*7C
$INHDG,132.9,0,E,1.0,E*58
$GNVTG,,T,,M,0.654,N,1.212,K,A*3A
$GNGGA,141948.00,5115.59932,N,00040.68149,W,1,04,3.84,58.8,M,46.2,M,,*6D
```

TackTick Display On:

```
$INHDG,132.7,0,E,1.0,E*56
$INHDG,132.8,0,E,1.0,E*59
$GNRMC,141600.00,A,5115.59594,N,00040.66776,W,0.542,,060323,,,A*7F
$GNVTG,,T,,M,0.542,N,1.005,K,A*3A
$GNGGA,141600.00,5115.59594,N,00040.66776,W,1,04,5.79,89.2,M,46.2,M,,*68
$INMWV,154.0,R,0.0,N,A*3a
```

### Emulate enabled:

TackTick Display Off:

```
$INHDG,132.8,0,E,1.0,E*59
$INHDG,132.9,0,E,1.0,E*58
$GNRMC,135912.00,A,5115.59615,N,00040.67671,W,0.373,,060323,,,A*79
$GNVTG,,T,,M,0.373,N,0.690,K,A*35
$GNGGA,135912.00,5115.59615,N,00040.67671,W,1,04,5.39,104.0,M,46.2,M,,*58
```

TackTick Display On:

```
$INHDG,132.9,0,E,1.0,E*58
$INVHW,133.8,T,132.8,M,0.3,N,,K*7e
$INMWV,154.0,R,0.0,N,A*3a
$INMWV,180.0,T,0.3,N,A*36
$INHDG,132.8,0,E,1.0,E*59
$GNRMC,140412.00,A,5115.61215,N,00040.68673,W,0.238,,060323,,,A*7A
$GNVTG,,T,,M,0.238,N,0.441,K,A*35
$GNGGA,140412.00,5115.61215,N,00040.68673,W,1,04,5.26,84.2,M,46.2,M,,*60
```

**Reply by Rodemfr** [reply]
Thanks for testing !
